src/commands/captureDiagnostics.ts

```diff
@@ -0,0 +1,209 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the VS Code Swift open source project
+//
+// Copyright (c) 2021-2024 the VS Code Swift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of VS Code Swift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import * as fs from "fs/promises";
+import * as path from "path";
+import * as vscode from "vscode";
+import { tmpdir } from "os";
+import { exec } from "child_process";
+import { Writable } from "stream";
+import { SwiftOutputChannel } from "../ui/SwiftOutputChannel";
+import { WorkspaceContext } from "../WorkspaceContext";
+import { Version } from "../utilities/version";
+import { execFileStreamOutput } from "../utilities/utilities";
+import configuration from "../configuration";
+
+export async function captureDiagnostics(ctx: WorkspaceContext) {
+    const diagnosticsDir = path.join(
+        tmpdir(),
+        `vscode-diagnostics-${formatDateString(new Date())}`
+    );
+
+    try {
+        await fs.mkdir(diagnosticsDir);
+        await writeLogFile(diagnosticsDir, "logs.txt", extensionLogs(ctx));
+        await writeLogFile(diagnosticsDir, "environment.txt", environmentLogs(ctx));
+
+        if (ctx.swiftVersion.isGreaterThanOrEqual(new Version(6, 0, 0))) {
+            // sourcekit-lsp diagnose command is only available in 6.0 and higher.
+            // await writeLogFile(diagnosticsDir, "sourcekit-lsp.txt", );
+            await sourcekitDiagnose(ctx, diagnosticsDir);
+        } else {
+            await writeLogFile(diagnosticsDir, "sourcekit-lsp.txt", sourceKitLogs(ctx));
+        }
+
+        await writeLogFile(diagnosticsDir, "diagnostics.txt", diagnosticLogs());
+
+        ctx.outputChannel.log(`Saved diagnostics to ${diagnosticsDir}`);
```

Comment:
I planned on adding this to the `Saved diagnostic logs` information message but it looks like VS Code forbids newlines in these messages, making it a bit awkward. https://github.com/Microsoft/vscode/blob/8616dbae8bc2abf7972a45449b0fb6b2b2d0f429/src/vs/workbench/common/notifications.ts#L412-L413

Reply:
Ah, I forgot that notifications can’t contain newlines. I hit this as well for indexing progress…

What do you think about adding the following sentence (or something of the like): 
> This folder contains information that helps developers of Swift for VS Code diagnose issues and may contain information about your source code and VS Code setup.

Reply:
I reworked the initial dialog to be a modal instead of an information panel. It gives a full accounting of what is captured, and offers two options, Full or Minimal diagnostic bundle. If the user chooses Minimal then sourcekit-lsp logs are omitted. I put the call to action in this dialog.

<img width="1139" alt="Screenshot 2024-06-11 at 12 09 58 PM" src="https://github.com/swiftlang/vscode-swift/assets/667875/ba24cdd2-9cec-4732-9e03-ae57e3a36c65">

Reply:
But even the minimal diagnostic bundle could include sensitive information, right? Error messages from the Problems pane alone might contain eg. strings from the user’s code base and those might be sensitive.